Dev/change timestamps

MauritsVM opened the pull request:
Implemented the code to change the timestamps of the LiDAR pointcloud

MauritsVM committed:
Corrected the LiDAR timestamps to be in line with the poses

wouter-heerwegh commented on src/dlio/odom.cc:
```diff
@@ -26,9 +26,9 @@ dlio::OdomNode::OdomNode(ros::NodeHandle node_handle) : nh(node_handle) {
+  this->lidar_sub = this->nh.subscribe("/ouster/points", 1,
       &dlio::OdomNode::callbackPointCloud, this, ros::TransportHints().tcpNoDelay());
-  this->imu_sub = this->nh.subscribe("imu", 1000,
+  this->imu_sub = this->nh.subscribe("/ouster/imu", 1000,
```
These changes remove the functionality to change the topic names through arguments in the launch file. These changes should be reverted.

wouter-heerwegh commented on src/dlio/odom.cc:
```diff
@@ -64,6 +64,7 @@ dlio::OdomNode::OdomNode(ros::NodeHandle node_handle) : nh(node_handle) {
   this->imu_meas.lin_accel[2] = 0.;
 
   this->imu_buffer.set_capacity(this->imu_buffer_size_);
+  this->pose_buffer.set_capacity(this->imu_buffer_size_);
```
Buffer size is currently 5000. A seperate ros parameter could be made to change the size of this buffer

wouter-heerwegh commented on src/dlio/odom.cc:
```diff
@@ -458,7 +507,8 @@ void dlio::OdomNode::publishCloud(pcl::PointCloud<PointType>::ConstPtr published
   // published deskewed cloud
   sensor_msgs::PointCloud2 deskewed_ros;
```
This is running in a seperate thread, what happens if the `lidarPose` updates before the stamp is computed, or the `scan_header_stamp` changes before publishing?

wouter-heerwegh commented on src/dlio/odom.cc:
```diff
@@ -447,6 +449,53 @@ void dlio::OdomNode::publishToROS(pcl::PointCloud<PointType>::ConstPtr published
+      best_cost = cost;
+      best_prev_pose = prev_pose;
+      best_curr_pose = curr_pose;
+    }
```
This assumption is correct if the closest pose is behind the `lidarPose`. If the pose is in front of `lidarPose`, the previous best pose will be the one before that, and thus be further than the next pose. You would need to catch at which point you would go from behind the `lidarPose` to in front of it, for instance by calculating the angle in the frame of the `lidarPose` between the pose and the coordinate system of the lidar pose.

wouter-heerwegh commented on src/dlio/odom.cc:
```diff
@@ -447,6 +449,53 @@ void dlio::OdomNode::publishToROS(pcl::PointCloud<PointType>::ConstPtr published
+
+  // Required variables
+  auto curr_pose = this->pose_buffer.begin(); // Current pose
+  auto prev_pose = this->pose_buffer.begin(); // Previous pose
```
No poses are ever popped from the buffer, this will currently grow till `this->imu_buffer_size_`, which is currently 5000

MauritsVM committed:
Added additional transform from baselink to lidar

wouter-heerwegh commented on src/dlio/odom.cc:
```diff
@@ -447,18 +449,78 @@ void dlio::OdomNode::publishToROS(pcl::PointCloud<PointType>::ConstPtr published
+  transform_matrix.block(0,0,3,3) = (this->tempPose2.q.normalized() * this->tempPose1.q.normalized().inverse()).normalized().toRotationMatrix();
+  transform_matrix(0,3) = this->tempPose2.p[0] - this->tempPose1.p[0];
+  transform_matrix(1,3) = this->lidarPose.p[1] - this->tempPose1.p[1];
+  transform_matrix(2,3) = this->lidarPose.p[2] - this->tempPose1.p[2];
```
I assume this is supposed to be all `tempPose2`? But it is unclear to me what you're doing here. `tempPose1` seems to be set at [line 615-616](https://github.com/Formula-Electric-Belgium/direct_lidar_inertial_odometry/blob/04da041cddc6894b16d42c35a1933e4fa11029cd/src/dlio/odom.cc#L615-L616), but the value is also `lidarPose`? So nothing happens because of this transform?

At the moment we have the following issue:
![image](https://github.com/Formula-Electric-Belgium/direct_lidar_inertial_odometry/assets/20578313/4cd40f39-7d0b-4bc4-9e45-111750c90a7b)
Black = `this->state`, green = `this->lidarPose`
The state is interpolated by calculations and new data from the IMU, but the eventual path will always be a bit off from the POV of the actual lidar position (this is to smooth out the path). So the transform we want to apply is one from the `lidarPose` to the `state`, so that the scan is seen in the pov of the `lidarPose` from the `state` position.

The calculation if this transform can also be done around [line 862](https://github.com/Formula-Electric-Belgium/direct_lidar_inertial_odometry/blob/04da041cddc6894b16d42c35a1933e4fa11029cd/src/dlio/odom.cc#L862C5-L862C5). This removes the need for `tempPose1` and `tempPose2` and the transform can be directly applied to the transform passed to the `publishToROS` function on [line 866](https://github.com/Formula-Electric-Belgium/direct_lidar_inertial_odometry/blob/04da041cddc6894b16d42c35a1933e4fa11029cd/src/dlio/odom.cc#L866).

wouter-heerwegh commented on src/dlio/odom.cc:
```diff
@@ -447,18 +449,78 @@ void dlio::OdomNode::publishToROS(pcl::PointCloud<PointType>::ConstPtr published
   sensor_msgs::PointCloud2 deskewed_ros;
   pcl::toROSMsg(*deskewed_scan_t_, deskewed_ros);
-  deskewed_ros.header.stamp = this->scan_header_stamp;
+  deskewed_ros.header.stamp = this->correctTimestamp(this->scan_header_stamp);
```
The scan can be given the stamp of the `this-state` pose

MauritsVM committed:
Simplified baselink to odom transform

wouter-heerwegh committed:
No more duplicate cones, but way higher covariances in slam used

Changes made are mainly to try and have constant values for tempPose1 and tempTime,
As they are copied to the `publishToRos` function.

wouter-heerwegh commented:
@MauritsVM seeing that the issues on the slam side are fixed now, can we try to find a final solution for the stamp syncing? I'm guessing the develop branch is not yet perfect, right?

wouter-heerwegh committed:
Update some dependencies

wouter-heerwegh committed:
Make nanoflann depend on the exported service message

wouter-heerwegh commented:
Hey @MauritsVM, could you finalize this solution so that it can be merged into develop and possible main? That way we can mark this as a "stable" version to fall back on.

Last we spoke on this topic, I think we concluded on taking the latest pose that is published and using this and it's timestamp to do the transformation for the deskewed pointcloud. Can you confirm or elaborate on the approach that we will take?

MauritsVM committed:
Cleanup of DLIO

MauritsVM committed:
Merge branch 'dev/change_timestamps' of github.com:Formula-Electric-Belgium/direct_lidar_inertial_odometry into dev/change_timestamps

MauritsVM reviewed (commented):
Removed unused correction functions

MauritsVM committed:
Removed unused functions and variables

wouter-heerwegh commented:
Looks good for the most part. I do think however that it could be possible for the published pose to be modified while copying data from and to it. It could be good to guard it with a lock, then copy the data that will be used and unlock.

Also I think the transformation might take long enough to cause the published pose to already be different when publishing the baselink frame, causing the frame to be different from the pose that is used for transformation. To fix this you could pass the copied data to the publish pointcloud function.

MauritsVM committed:
Added synchronisation

wouter-heerwegh commented on src/dlio/odom.cc:
```diff
@@ -26,9 +26,9 @@ dlio::OdomNode::OdomNode(ros::NodeHandle node_handle) : nh(node_handle) {
   this->deskew_size = 0;
 
-  this->lidar_sub = this->nh.subscribe("pointcloud", 1,
+  this->lidar_sub = this->nh.subscribe("/ouster/points", 1,
```
This should be changed back, you can just use 

`roslaunch direct_lidar_inertial_odometry dlio.launch pointcloud_topic:=/ouster/points imu_topic:=/ouster/imu`

This allows for maximum flexibility. You could also incloude the launch file in one of the launch files in the launch package and pass the `pointcloud_topic` and `imu_topic` as arguments.

MauritsVM replied:
I always forget about that, my computer crashes almost every catkin build and having to type all the arguments grows very tiresome so I just change that every time before building so it's not necessary but I sometimes forget to change it back before pushing

wouter-heerwegh commented on src/dlio/odom.cc:
```diff
@@ -385,14 +392,16 @@ void dlio::OdomNode::publishToROS(pcl::PointCloud<PointType>::ConstPtr published
+  publishedPoseMtx.lock();
+  transformStamped.transform.translation.x = this->publishedPose.p[0];
+  transformStamped.transform.translation.y = this->publishedPose.p[1];
+  transformStamped.transform.translation.z = this->publishedPose.p[2];
```
The data used here could still be different from the poseCopy variable, resulting in a mismatch

wouter-heerwegh reviewed (changes requested):
Solution looks good, except for `poseCopy`. I would move the copy of the pose outside `publishPointcloud`, because the baselink frame is not using the copied pose and stamp. Since it comes after the publish of the pointcloud, chances are high that the pose has changed.

MauritsVM committed:
Adjusted sync for better timing

MauritsVM committed:
Removed test frame
